Break complex `Reduce` operators into simpler ones

frankmcsherry opened the pull request:
This is an experimental PR where we break apart `Reduce` operators that would be rendered with the `Collation` plan into atomic `Reduce` operators that are joined together instead. This has the potential to be (much) better than the collation plan, or .. to be worse. If we put an `ArrangeBy` at the end it wouldn't be worse, but it could be much better.

This PR is mostly to look at some plans and see what changes and how.

Edit: More explanation here: https://materializeinc.slack.com/archives/C02PPB50ZHS/p1673028168703109

Edit2: Slack message copy/pasted for visibility:

> Other Reduce thought: We have various flavors of reduce plans, roughly three (accumulable, hierarchical, and "generic"). We collate these together with another reduce though .. we should just use a delta join. All of the constituents present arrangements of their outputs, and a delta join would use no additional memory (unlike the collation operator). Moreover, we could then push down predicates and projection and mapping and such.
> Downside: delta joins don't produce an output arrangement, so it wouldn't always be the case that Reduce has an arrangement of its output. We could always make one at not much additional cost (and I think strictly less than the collation operator).

Edit 3: scope has changed to only breaking apart enough aggregates to prevent collation. Fixes https://github.com/MaterializeInc/database-issues/issues/2273

### Motivation

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

vmarcos commented on test/sqllogictest/transform/aggregation_nullability.slt:
```diff
@@ -193,26 +251,55 @@ EXPLAIN WITH(arity, join_impls) select t1.f1, count(t2.f1), sum(t2.f1), max(t2.f
+            %2 » %4[#0]UKAif » %0[#0]UKA » %1[#0]UKA » %3[#0]UKA
+            %3 » %4[#0]UKAif » %0[#0]UKA » %1[#0]UKA » %2[#0]UKA
+            %4 » %0[#0]UKA » %1[#0]UKA » %2[#0]UKA » %3[#0]UKA
+          ArrangeBy keys=[[#0]] // { arity: 2 }
+            Reduce group_by=[#0] aggregates=[max(#1)] // { arity: 2 }
+              Get l2 // { arity: 2 }
+          ArrangeBy keys=[[#0]] // { arity: 2 }
+            Reduce group_by=[#0] aggregates=[min(#1)] // { arity: 2 }
+              Get l2 // { arity: 2 }
+          ArrangeBy keys=[[#0]] // { arity: 2 }
+            Reduce group_by=[#0] aggregates=[min(#1)] // { arity: 2 }
+              Get l3 // { arity: 2 }
+          ArrangeBy keys=[[#0]] // { arity: 2 }
+            Reduce group_by=[#0] aggregates=[max(#1)] // { arity: 2 }
+              Get l3 // { arity: 2 }
+          ArrangeBy keys=[[#0]] // { arity: 5 }
+            Reduce group_by=[#0] aggregates=[count(#2), sum(#2), count(#1), sum(#1)] // { arity: 5 }
+              Get l1 // { arity: 3 }
```
This is an interesting case in that it would not be a bad idea to split the hierarchical aggregates that operate on different attributes, but it seems to be still better to compute min/max on the same attribute in a single aggregation as you pointed out in [Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1673109203160559).

vmarcos reviewed (approved):
The shape of the resulting plans looks pretty good to me, where the style starts resembling a bit how planning could look like under vertical partitioning. I am not sure about splitting and rejoining min/max aggregates on the same attributes, though. So most of my questions relate to the latter point.

frankmcsherry commented on test/sqllogictest/autogenerated/all_parts_essential.slt:
```diff
@@ -358,22 +377,36 @@ GROUP BY 1,
-            Get materialize.public.customer // { arity: 8 }
+  Return // { arity: 4 }
+    Project (#0..=#2, #5) // { arity: 4 }
+      Join on=(#0 = #3 AND #1 = #4) type=differential // { arity: 6 }
+        implementation
+          %1[#0, #1] » %0[#0, #1]UKKA
+        ArrangeBy keys=[[#0, #1]] // { arity: 3 }
+          Reduce group_by=[(#1 * #2), #3] aggregates=[max(#0)] // { arity: 3 }
+            Project (#1..=#4) // { arity: 4 }
+              Get l0 // { arity: 5 }
+        ArrangeBy keys=[[#0, #1]] // { arity: 3 }
+          Reduce group_by=[(#1 * #2), #3] aggregates=[max(#0)] // { arity: 3 }
+            Project (#0, #2..=#4) // { arity: 4 }
+              Get l0 // { arity: 5 }
```
Yes, and it's actually intentional. The code used to do this but a design regression lost that ability at one point. The reason you might want this is that if you are doing
```sql
SELECT
    key,
    MAX(height),
    MAX(weight),
    MAX(age),
    MAX(salary),
...
GROUP BY key
```
then with the join plan you keep a footprint proportional to the sum of distinct values for each of the attributes. With the non-join plan, as we currently do it, you keep a footprint proportional to the distinct products of values, which can be substantially larger (and was for the only user we ever asked).

Though, 100% doing this intentionally makes sense, but I think the join plan reduces the worst case bound, at the potential cost of the best case bound (maybe `height` and `weight` are correlated, and there aren't that many distinct pairs?).

vmarcos replied:
That's an interesting point, and I am definitely all for ensuring good worst-case behavior, even if it costs us a bit on the best case. Does it make sense here to make a distinction between monotonic vs. non-monotonic plans? Perhaps I am missing something, but the monotonic rendering seems to be better behaved wrt. the effect you describe?

frankmcsherry commented on test/sqllogictest/explain/physical_plan_as_json.slt:
```diff
@@ -3695,748 +3564,3105 @@ GROUP BY a
 EOF
 
-# Test Reduce::Collated (global aggregate).
+# Test Reduce::Collated (with GROUP BY).
```
Tbh, I have no idea what this text is and did not intentionally change it.

vmarcos commented:
For what it is worth, with this PR, we get on a query adapted from CH-benCHmark's Q1:
```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_discount) as max_disc,
         min(l_discount) as min_disc,
         max(l_tax) as max_tax,
         min(l_tax) as min_tax
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |   sum_price    |      avg_qty       |     avg_price      | count_order | max_disc | min_disc | max_tax | min_tax 
--------------+----------+----------------+--------------------+--------------------+-------------+----------+----------+---------+---------
            1 | 38248246 | 57357083080.11 | 25.498830666666667 |     38238.05538674 |     1500000 |      0.1 |        0 |    0.08 |       0
            2 | 32789215 | 49179568260.61 | 25.500467403105237 | 38247.392544422735 |     1285828 |      0.1 |        0 |    0.08 |       0
            3 | 27349884 | 41020910796.75 |  25.52738208352856 |  38287.41881768052 |     1071394 |      0.1 |        0 |    0.08 |       0
            4 | 21857330 |  32777528689.6 | 25.504022683383603 | 38246.155189349076 |      857015 |      0.1 |        0 |    0.08 |       0
            5 | 16411322 | 24616149732.39 | 25.511664311574773 | 38266.201139444755 |      643287 |      0.1 |        0 |    0.08 |       0
            6 | 10937358 | 16400552024.22 | 25.490847647237047 |  38223.48806539725 |      429070 |      0.1 |        0 |    0.08 |       0
            7 |  5485440 |  8225518317.52 | 25.558729108521533 | 38325.785070053724 |      214621 |      0.1 |        0 |    0.08 |       0
(7 rows)

Time: 2816.371 ms (00:02.816)
```
While this is not a highly controlled run, on main, performance appears to be better:
```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_discount) as max_disc,
         min(l_discount) as min_disc,
         max(l_tax) as max_tax,
         min(l_tax) as min_tax
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |   sum_price    |      avg_qty       |     avg_price      | count_order | max_disc | min_disc | max_tax | min_tax 
--------------+----------+----------------+--------------------+--------------------+-------------+----------+----------+---------+---------
            1 | 38248246 | 57357083080.11 | 25.498830666666667 |     38238.05538674 |     1500000 |      0.1 |        0 |    0.08 |       0
            2 | 32789215 | 49179568260.61 | 25.500467403105237 | 38247.392544422735 |     1285828 |      0.1 |        0 |    0.08 |       0
            3 | 27349884 | 41020910796.75 |  25.52738208352856 |  38287.41881768052 |     1071394 |      0.1 |        0 |    0.08 |       0
            4 | 21857330 |  32777528689.6 | 25.504022683383603 | 38246.155189349076 |      857015 |      0.1 |        0 |    0.08 |       0
            5 | 16411322 | 24616149732.39 | 25.511664311574773 | 38266.201139444755 |      643287 |      0.1 |        0 |    0.08 |       0
            6 | 10937358 | 16400552024.22 | 25.490847647237047 |  38223.48806539725 |      429070 |      0.1 |        0 |    0.08 |       0
            7 |  5485440 |  8225518317.52 | 25.558729108521533 | 38325.785070053724 |      214621 |      0.1 |        0 |    0.08 |       0
(7 rows)

Time: 1934.263 ms (00:01.934)
```

vmarcos commented:
The difference drops somewhat when we combine with only one hierarchical aggregate instead; however, it is still favorable to main. With this PR:
```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_discount) as max_disc
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |   sum_price    |      avg_qty       |     avg_price      | count_order | max_disc 
--------------+----------+----------------+--------------------+--------------------+-------------+----------
            1 | 38248246 | 57357083080.11 | 25.498830666666667 |     38238.05538674 |     1500000 |      0.1
            2 | 32789215 | 49179568260.61 | 25.500467403105237 | 38247.392544422735 |     1285828 |      0.1
            3 | 27349884 | 41020910796.75 |  25.52738208352856 |  38287.41881768052 |     1071394 |      0.1
            4 | 21857330 |  32777528689.6 | 25.504022683383603 | 38246.155189349076 |      857015 |      0.1
            5 | 16411322 | 24616149732.39 | 25.511664311574773 | 38266.201139444755 |      643287 |      0.1
            6 | 10937358 | 16400552024.22 | 25.490847647237047 |  38223.48806539725 |      429070 |      0.1
            7 |  5485440 |  8225518317.52 | 25.558729108521533 | 38325.785070053724 |      214621 |      0.1
(7 rows)

Time: 1627.088 ms (00:01.627)
```
And on main:
```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_discount) as max_disc
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |   sum_price    |      avg_qty       |     avg_price      | count_order | max_disc 
--------------+----------+----------------+--------------------+--------------------+-------------+----------
            1 | 38248246 | 57357083080.11 | 25.498830666666667 |     38238.05538674 |     1500000 |      0.1
            2 | 32789215 | 49179568260.61 | 25.500467403105237 | 38247.392544422735 |     1285828 |      0.1
            3 | 27349884 | 41020910796.75 |  25.52738208352856 |  38287.41881768052 |     1071394 |      0.1
            4 | 21857330 |  32777528689.6 | 25.504022683383603 | 38246.155189349076 |      857015 |      0.1
            5 | 16411322 | 24616149732.39 | 25.511664311574773 | 38266.201139444755 |      643287 |      0.1
            6 | 10937358 | 16400552024.22 | 25.490847647237047 |  38223.48806539725 |      429070 |      0.1
            7 |  5485440 |  8225518317.52 | 25.558729108521533 | 38325.785070053724 |      214621 |      0.1
(7 rows)

Time: 1379.412 ms (00:01.379)
```

vmarcos commented:
I am trying to get a better feel for the performance characteristics of this PR by running variations of the first query above adapted from CH-benCHmark's Q1. For full disclosure, the numbers in my previous comments were generated with a monotonic source, so this means that we were in the happy case of aggregate computation. All numbers were with 1 worker in my local development environment and I created all PK and FK indexes on the TPC-H schema with a database at scale factor 1.

With a non-monotonic source, we get higher absolute query processing times, though main still performs better. With this PR:

```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_discount) as max_disc,
         min(l_discount) as min_disc,
         max(l_tax) as max_tax,
         min(l_tax) as min_tax
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |  sum_price  |      avg_qty       |     avg_price      | count_order | max_disc | min_disc | max_tax | min_tax 
--------------+----------+-------------+--------------------+--------------------+-------------+----------+----------+---------+---------
            1 | 38255417 | 57347642187 | 25.503611333333332 |       38231.761458 |     1500000 |     0.08 |        0 |     0.1 |       0
            2 | 32744738 | 49079433329 |  25.46966056515175 |  38175.18734215023 |     1285637 |     0.08 |        0 |     0.1 |       0
            3 | 27341357 | 40979779888 | 25.496839164334546 |  38215.17918054204 |     1072343 |     0.08 |        0 |     0.1 |       0
            4 | 21874643 | 32784671314 |  25.49426821223698 | 38209.592893888504 |      858022 |     0.08 |        0 |     0.1 |       0
            5 | 16393227 | 24574702761 | 25.468292857231635 |  38178.92150194509 |      643672 |     0.08 |        0 |     0.1 |       0
            6 | 10950676 | 16420605972 | 25.504292375269582 | 38243.843182739205 |      429366 |     0.08 |        0 |     0.1 |       0
            7 |  5484325 |  8220701677 | 25.524990575302173 |  38260.55764889859 |      214861 |     0.08 |        0 |     0.1 |       0
(7 rows)

Time: 21308.044 ms (00:21.308)

```

On main:

```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_discount) as max_disc,
         min(l_discount) as min_disc,
         max(l_tax) as max_tax,
         min(l_tax) as min_tax
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |  sum_price  |      avg_qty       |     avg_price      | count_order | max_disc | min_disc | max_tax | min_tax 
--------------+----------+-------------+--------------------+--------------------+-------------+----------+----------+---------+---------
            1 | 38255417 | 57347642187 | 25.503611333333332 |       38231.761458 |     1500000 |     0.08 |        0 |     0.1 |       0
            2 | 32744738 | 49079433329 |  25.46966056515175 |  38175.18734215023 |     1285637 |     0.08 |        0 |     0.1 |       0
            3 | 27341357 | 40979779888 | 25.496839164334546 |  38215.17918054204 |     1072343 |     0.08 |        0 |     0.1 |       0
            4 | 21874643 | 32784671314 |  25.49426821223698 | 38209.592893888504 |      858022 |     0.08 |        0 |     0.1 |       0
            5 | 16393227 | 24574702761 | 25.468292857231635 |  38178.92150194509 |      643672 |     0.08 |        0 |     0.1 |       0
            6 | 10950676 | 16420605972 | 25.504292375269582 | 38243.843182739205 |      429366 |     0.08 |        0 |     0.1 |       0
            7 |  5484325 |  8220701677 | 25.524990575302173 |  38260.55764889859 |      214861 |     0.08 |        0 |     0.1 |       0
(7 rows)

Time: 12227.115 ms (00:12.227)
```

vmarcos commented:
Now, let's change the query to manifest the effect pointed out in https://github.com/MaterializeInc/materialize/pull/17013#discussion_r1066099381, again with a non-monotonic source. The domains of `l_discount` and `l_tax` are pretty small; let's use instead `l_orderkey` and `l_partkey`.

With this PR:

```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_orderkey) as max_order,
         min(l_orderkey) as min_order,
         max(l_partkey) as max_part,
         min(l_partkey) as min_part
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |  sum_price  |      avg_qty       |     avg_price      | count_order | max_order | min_order | max_part | min_part 
--------------+----------+-------------+--------------------+--------------------+-------------+-----------+-----------+----------+----------
            1 | 38255417 | 57347642187 | 25.503611333333332 |       38231.761458 |     1500000 |   6000000 |         1 |   200000 |        1
            2 | 32744738 | 49079433329 |  25.46966056515175 |  38175.18734215023 |     1285637 |   6000000 |         1 |   200000 |        1
            3 | 27341357 | 40979779888 | 25.496839164334546 |  38215.17918054204 |     1072343 |   6000000 |         1 |   200000 |        1
            4 | 21874643 | 32784671314 |  25.49426821223698 | 38209.592893888504 |      858022 |   6000000 |         2 |   200000 |        1
            5 | 16393227 | 24574702761 | 25.468292857231635 |  38178.92150194509 |      643672 |   6000000 |         2 |   200000 |        1
            6 | 10950676 | 16420605972 | 25.504292375269582 | 38243.843182739205 |      429366 |   6000000 |         2 |   200000 |        1
            7 |  5484325 |  8220701677 | 25.524990575302173 |  38260.55764889859 |      214861 |   5999971 |        64 |   200000 |        1
(7 rows)

Time: 82687.735 ms (01:22.688)

```

With main:

```sql
materialize=> select  l_linenumber,
         sum(l_quantity) as sum_qty,
         sum(l_extendedprice) as sum_price,
         avg(l_quantity) as avg_qty,
         avg(l_extendedprice) as avg_price,
         count(*) as count_order,
         max(l_orderkey) as max_order,
         min(l_orderkey) as min_order,
         max(l_partkey) as max_part,
         min(l_partkey) as min_part
from     lineitem
group by l_linenumber 
order by l_linenumber;
 l_linenumber | sum_qty  |  sum_price  |      avg_qty       |     avg_price      | count_order | max_order | min_order | max_part | min_part 
--------------+----------+-------------+--------------------+--------------------+-------------+-----------+-----------+----------+----------
            1 | 38255417 | 57347642187 | 25.503611333333332 |       38231.761458 |     1500000 |   6000000 |         1 |   200000 |        1
            2 | 32744738 | 49079433329 |  25.46966056515175 |  38175.18734215023 |     1285637 |   6000000 |         1 |   200000 |        1
            3 | 27341357 | 40979779888 | 25.496839164334546 |  38215.17918054204 |     1072343 |   6000000 |         1 |   200000 |        1
            4 | 21874643 | 32784671314 |  25.49426821223698 | 38209.592893888504 |      858022 |   6000000 |         2 |   200000 |        1
            5 | 16393227 | 24574702761 | 25.468292857231635 |  38178.92150194509 |      643672 |   6000000 |         2 |   200000 |        1
            6 | 10950676 | 16420605972 | 25.504292375269582 | 38243.843182739205 |      429366 |   6000000 |         2 |   200000 |        1
            7 |  5484325 |  8220701677 | 25.524990575302173 |  38260.55764889859 |      214861 |   5999971 |        64 |   200000 |        1
(7 rows)

Time: 39159.046 ms (00:39.159)
```

I expected to see increased query processing times; however, we get the same with this PR and still consistently close to 2x slower, which is pretty puzzling.

To understand a bit better, I tried to create the query above using `l_orderkey` and `l_partkey` as an indexed view and check how many records the dataflow is keeping around. With main, we get 30,949,728 records, while with this PR, we get 73,246,205 records.

From what I can tell, the effect is coming from our hierarchical reduces being forced to maintain so many records. Since we are vertically partitioning hierarchical aggregates in this PR, we are increasing the number of arrangement hierarchies. Perhaps one less bold step would be to only break down reductions by type (more in line with a reduce collation) and not by component columns? It would still not get rid of the effect pointed out in https://github.com/MaterializeInc/materialize/pull/17013#discussion_r1066099381, but if IIUC this would be enough for us to get rid of the reduce collation rendering?

frankmcsherry commented:
The reason is that in that query the number of distinct values for `l_discount` and `l_tax` is relatively small, and the number of groups is also small. There is no harm in doing the primitive aggregation done on `main`. I suspect if you try it with more heterogeneity in each of a larger number of groups you may see different results. For example, consider data with columns `(ZIP, height, weight, age)`, a cross product of all reasonable heights and weights and ages in each group, and then query
```
SELECT ZIP
    MAX(height),
    MAX(weight),
    MAX(age),
FROM that_data
GROUP BY ZIP;
```
The memory footprint of this should be proportional to the sum of distinct (ZIP, height), distinct (ZIP, weight), and distinct (ZIP, age) records, rather than the distinct `(ZIP, height, weight, age)` records.

frankmcsherry commented:
To do a thing I should have done with the PR, here is a scenario that demonstrates the difference between the approaches. It is meant to reveal the difference between the things, and I hope I have not screwed it up. :)

First, let's create some maximally heterogeneous data, unioning in a table to make sure the data aren't closed out and we can see the consequences of their dataflows once live.
```sql
CREATE TABLE foo (a int, b int, c int, d int);

CREATE MATERIALIZED VIEW test AS 
SELECT * FROM 
    generate_series(1, 100) as a, 
    generate_series(1, 100) as b, 
    generate_series(1, 100) as c, 
    generate_series(1, 100) as d
UNION ALL
SELECT * FROM foo;
```
At this point, on `main`, we can do two styles of queries that determine the same thing.
```sql
CREATE MATERIALIZED VIEW max_old AS
SELECT a, MAX(b) as b, MAX(c) as c, MAX(d) as d
FROM test
GROUP BY a;
```
I wanted this to work, but I bailed out at 100GB of memory on my laptop (after many minutes).
```sql
CREATE MATERIALIZED VIEW max_new AS
WITH
    b_max as (SELECT a, MAX(b) as b from test GROUP BY a),
    c_max as (SELECT a, MAX(c) as c from test GROUP BY a),
    d_max as (SELECT a, MAX(d) as d from test GROUP BY a)
SELECT b_max.a, b, c, d
FROM
    b_max, c_max, d_max
WHERE
    b_max.a = c_max.a AND b_max.a = d_max.a
;
```
This is currently being maintained on my laptop in 243MB of memory, some 100MB of which was already part of `clusterd` when I started.

frankmcsherry commented:
I removed the `d` column (bringing the records down to 1M total) and just the query times are already different:
```
max_old: Time: 7132.956 ms (00:07.133)
max_new: Time: 2313.396 ms (00:02.313)
```
The memory cost of each as a materialized view is imperceptible for `max_new` (from 204MB to 206MB), and over a GB for `max_old` (from 206MB to 1.24GB).

vmarcos commented:
Ah, your comments clarify things a lot, thanks! My second set of measurements above did not manifest the effect because `(l_orderkey, linenumber)` is a key for `lineitem`. So this functional dependency made the size of the relation be not too far from the sums of the sizes of the projections.

Your explanation and scenario make it clear that this PR is extremely desirable to avoid terrible worst-case performance, even if it costs us some regressions in other better behaved cases.

I have a few follow-up questions, also formulated as you did on `main`:

1) Consider computing `MAX`/`MIN` on the same attribute. In such a case, this PR seems to vertically partition more than desirable? For example, consider:

```sql
CREATE MATERIALIZED VIEW maxmin_old AS
SELECT a, MAX(b) as max_b, MIN(b) as min_b
FROM test
GROUP BY a;
```

and

```sql
CREATE MATERIALIZED VIEW maxmin_new AS
WITH
    b_max as (SELECT a, MAX(b) as max_b from test GROUP BY a),
    b_min as (SELECT a, MIN(b) as min_b from test GROUP BY a)
SELECT b_max.a, max_b, min_b
FROM
    b_max, b_min
WHERE
    b_max.a = b_min.a;
```

In `maxmin_new`, we maintain two stacks of arrangements when one would have been sufficient. I get 93,200 dataflow records for `maxmin_old` and 180,200 for `maxmin_new`. Query processing times also differ by a bit (573.509 ms vs. 629.909 ms in my machine).

2) The PR does not currently consider the `monotonic` flag. Would it be desirable to vertically partition the hierarchical aggregates under monotonic rendering? I put the contents of `test` without column `d` into a Kafka topic and created a corresponding source and view:

```sql
CREATE CONNECTION kafka FOR KAFKA BROKER 'localhost:9092';

CREATE SOURCE test_src
FROM  KAFKA CONNECTION kafka (TOPIC 'test')
FORMAT CSV
WITH 3 COLUMNS
DELIMITED BY '  '
WITH (SIZE = '1');

CREATE VIEW test_m AS
SELECT
    column1::integer as a,
    column2::integer as b,
    column3::integer as c
FROM test_src;
```

Revising your materialized view definitions to work on `test_m` and without the `d` column, we get about 200 dataflow records maintained under monotonic rendering for `max_old` and 400 for `max_new`, and the query processing times are also better for the SQL in `max_old` (roughly 1411.590 ms vs 2278.332 ms in my machine).

Given that this PR is fixing a huge potential problem, I would not oppose merging it without addressing the points above (if they would make sense to address anyhow).

frankmcsherry commented:
My sense at the moment is that this perhaps all belongs in LIR planning for `Reduce`, as many of the nits and curious bits relate to things like monotonicity discovered late in the process, or quirks of how we end up rendering things (e.g. that `MinMax` could be fused together, but they shouldn't be fused with `JsonbAggr` even if applied to the same `expr`, although `JsonbAgg` could be fused with `ListConcat` because they have the same strategy, etc... We also need to mind whether the `distinct` bool is set, stuff like that.

vmarcos commented:
I see... I'd defer here to someone with more optimizer ownership, perhaps @ggevay. The discussion so far indicates that some low-level planning concerns (types of aggregates, monotonicity) can help us make a better determination of situations where this transform could degrade performance (by a small factor). So it seems better suited for a planning pass closer to rendering than at the logical level. One way or another, I am favorable to having it in place since it can avoid very large performance degradations in some instances.

ggevay commented:
If monotonicity info is important, then we don't have much choice: we indeed have to put it in the LIR planning, as @frankmcsherry said above.

frankmcsherry commented:
Although I don't disagree about putting it in LIR, I think the thread above tells us that we can certainly make more informed decisions if the call gets made with information like unique key structure to informs us about heterogeneity (or other proxies for stats). LIR seems super-safe for the  moment, though.

ggevay commented:
> we can certainly make more informed decisions if the call gets made with information like unique key structure

We do have unique key info during the LIR planning, because we are looking at an MIR plan, so we can just call `typ`, right? Or am I misunderstanding something? (Is the quoted sentence not meant to be a counter-argument against putting it in LIR planning?)

frankmcsherry commented:
Historically, LIR has been a translation of things expressible in MIR into more specific implementation details. Given that the whole of this transform *can* be expressed in MIR, and given that MIR is where we have the type information like keys, MIR seems like a more natural home for the "optimization". LIR seems a bit like a (fine for now) place of last resort, as it isn't meant to be a place for further optimizations.

frankmcsherry commented:
Sorry, just to restate a take: MIR is a fine place to end up having this, as for example MIR *is* where we do monotonicity analysis, we just do it as part of the whole dataflow analysis as it cannot be done without sources or `as_of`/`until` information. It suggests that just having this as a transformation isn't great, as we usually apply those *before* we do whole dataflow optimization. But even that happens before we lower to LIR .

mgree commented:
I am leaving this comment to remind myself that Frank and I talked about this as a possible place to apply cost-based decisions.

vmarcos commented:
Reflecting on the discussion in this PR, I see the approach here tackling simultaneously two different problems:
1. `ReducePlan::Collation` takes up more memory than necessary (solvable by using a delta join instead).
2. `HierarchicalPlan::Bucketed` suffers when multiple uncorrelated aggregates are computed (solvable by vertical partitioning of hierarchical aggregates when neither monotonic nor having a strong correlation signal, e.g., FDs).

We can try to tackle both problems at once, as in this PR, or separate the work out on each. There might be an advantage to the latter in that: (a) A solution to problem 1 could be executed at MIR-level and a solution to problem 2 at LIR-level; (b) While both problems are important, problem 2 seems to be more serious than problem 1.

To support prioritization, I collected some data and wrote down observations in [an internal document](https://www.notion.so/materialize/Production-Analysis-of-MIN-MAX-Aggregations-347de8b1193741dda307aeff3a4bb9bf).

frankmcsherry commented on test/sqllogictest/transform/aggregation_nullability.slt:
```diff
@@ -158,25 +168,35 @@ EXPLAIN WITH(arity, join implementations) select t1.f1, count(t2.f1), sum(t2.f1)
-          Project (#0, #1, #0) // { arity: 3 }
-            Get l1 // { arity: 2 }
+    Project (#0..=#2, #6, #7, #3, #4, #8, #9) // { arity: 9 }
+      Filter (#4 >= 0) // { arity: 10 }
```
It looks like this should be able to be pushed down to the first `Reduce`, essentially becoming `HAVING SUM(#1) >= 0`.

ggevay replied:
It's not pushed in MIR because that would prevent arrangement re-use, so `JoinImplementation` actually lifts it away, I guess.

(But I guess it's pushed during the mir-to-lir lowering into the closure that is _after_ the corresponding join stage.)

frankmcsherry replied:
Ok, it feels like there's maybe a tension that we want the arrangement output from `Reduce`, but also we know how to push filters into `Reduce`; does the LIR translation do that level of MFP pushdown (through the entire `Join` operator, and on to the next operator). It did not when I originally wrote it, iirc, and if that's still the case perhaps we want to think about pushing it down despite the join arrangement interactions.

ggevay replied:
No, the filter will currently end up inside the LIR Join node.

I guess we could add a special case in `JoinImplementation`: If an already-arranged input is a Reduce, then there is no need to lift away the Filter, because arrangement reuse will happen even with the Filter being left there, because lowering will push the Filter into the `mfp_after` of the `Reduce`. Do I understand this correctly?

frankmcsherry replied:
I think the only case we don't want to push down the filter is when it is a `Get` operator with an arrangement it references? If it is a bespoke `ArrangeBy` .. we want to push down the filter. If it is a `Reduce` we should push it down. If it is a `Threshold` or a `TopK`, .. I don't think we make those arrangements available .. but if we did we could push down a filter to them.

frankmcsherry commented:
Rebased and rewritten to only shred `Reduce` enough to avoid the `ReducePlan::Collation` variant. This means that e.g. mins and maxes remain in the same `Reduce`, and all the various `_agg` things also get one `Reduce`. This *should* be pretty close to regression free, in the sense that we would have rendered them as this, followed by a DD `reduce` to collate the results. Instead we use a `Join` which avoids two arrangements.

Still lots of opportunities for regressions, of course. There seem to be several filters not being pushed down, for example.

frankmcsherry commented on test/sqllogictest/not-null-propagation.slt:
```diff
@@ -321,19 +321,30 @@ Explained Query:
         Union // { types: "(integer?, integer?, bigint?, bigint, numeric?, numeric?, bigint, integer list?)" }
-          Get l0 // { types: "(integer, integer, bigint, bigint, numeric, numeric, bigint, integer list)" }
+          Project (#5, #6, #0..=#4, #7) // { types: "(integer, integer, bigint, bigint, numeric, numeric, bigint, integer list)" }
+            Get l1 // { types: "(bigint, bigint, numeric, numeric, bigint, integer, integer, integer list)" }
           Map (null, null, null, 0, null, null, 0, null) // { types: "(integer?, integer?, bigint?, bigint, numeric?, numeric?, bigint, integer list?)" }
             Union // { types: "()" }
               Negate // { types: "()" }
                 Project () // { types: "()" }
-                  Get l0 // { types: "(integer, integer, bigint, bigint, numeric, numeric, bigint, integer list)" }
+                  Get l1 // { types: "(bigint, bigint, numeric, numeric, bigint, integer, integer, integer list)" }
```
This looks like a projection pushdown failure. Or .. given that it only happens later in the pipeline, and apparent regression that gets fixed by projection pushdown potentially. The new `Project (#5 ... ` stuff is just re-ordering the aggregate fields from the `l1 = CrossJoin .. ` term (and potentially projecting away duplicate keys, but it is a cross join so .. no keys).

ggevay replied:
`ProjectionPushdown` is able to push reorderings through some operators, but not through others. Specifically, it doesn't do it for `Get`s, because `gets` keeps a `BTreeSet` of columns for each id, rather than some other data structure that would be able to represent a reordering.

If it's desirable to push down reorderings, I can rework how `ProjectionPushdown` behaves with respect to reordering as part of fixing https://github.com/MaterializeInc/materialize/issues/17522 (which will involve a pretty big refactor of `ProjectionPushdown` anyway). But my feeling is that it's generally pretty hard to tell where a reordering should optimally be, because this is somewhat influenced by which operator is able to absorb it later during lowering.

antiguru commented on src/transform/src/reduce_reduction.rs:
```diff
@@ -0,0 +1,110 @@
+
+            // For each type of aggregation we'll plan the corresponding `Reduce`,
+            // and then join the at-least-two `Reduce` stages together.
+            // TODO: Perhaps we should introduce a `Let` stage rather than clone the input?
```
Re. your TODO: What do we do elsewhere? It seems better to introduce a `Let` stage rather than computing the same result again, or can we rely on subsequent optimization passes to clean this up for us?

ggevay replied:
`RelationCSE` is run after this pretty soon in the optimizer pipeline, and it should take care of this. Looking at the plans, it indeed does seem to take care of this.

antiguru commented on test/sqllogictest/attributes/mir_column_types.slt:
```diff
@@ -133,31 +140,41 @@ EXPLAIN OPTIMIZED PLAN WITH(types) AS TEXT FOR
+                ArrangeBy keys=[[#0]] // { types: "(double precision, bigint)" }
+                  Reduce group_by=[#0] aggregates=[count(*)] // { types: "(double precision, bigint)" }
+                    Project (#0) // { types: "(double precision)" }
+                      Get l0 // { types: "(double precision, boolean?)" }
+                ArrangeBy keys=[[#0]] // { types: "(double precision, boolean?)" }
+                  Reduce group_by=[#0] aggregates=[min(#1)] // { types: "(double precision, boolean?)" }
+                    Get l0 // { types: "(double precision, boolean?)" }
```
It seems we extract the input to a common cte!

antiguru commented on test/sqllogictest/transform/aggregation_nullability.slt:
```diff
@@ -864,35 +1025,43 @@ WHERE t1.f2 = derived.agg2;
   Return // { arity: 1 }
     Project (#1) // { arity: 1 }
-      Join on=(#0 = #2) type=differential // { arity: 3 }
+      Join on=(#0 = #2) type=delta // { arity: 3 }
```
This looks like an improvement as we're switching from differential to delta joins?

ggevay replied:
The join had 2 inputs before, and binary joins are always forced to be Differential joins.

frankmcsherry replied:
So actually, maybe a regression, because we went from a tree of two binary joins to one delta join, and that isn't *always* an improvement (delta joins don't have the zig-zag join property of quickly skipping ranges of key non-matches).

ggevay replied:
You mean there is a possibility of this being a regression, but not very likely, right?

This is one example of why we'd need a feature flag for this new transform.

frankmcsherry replied:
No idea if it will be a regression! Any change to MIR is 50-50 it will cause a regression, unless there is a reason it won't. If, for example, we capped the join we introduced here with an `ArrangeBy` I would say this couldn't be a regression. But .. join fusion results in regressions, so could totally be a regression even though locally everything is better.

antiguru reviewed (approved):
I think this looks fine. One part I'd like to know more about is the worst case scenario where the aggregate is small, but the key is larger. In this case, we might see a regression because we're arranging multiple times by the same key in different arrangements. How much of a risk is this?

I'm OK with merging the PR as-is given the potential positive impact due to leaner reductions.

ggevay reviewed (commented):
I'll continue reviewing a bit later.

I think a feature flag for this would be great.

frankmcsherry commented:
> One part I'd like to know more about is the worst case scenario where the aggregate is small, but the key is larger. In this case, we might see a regression because we're arranging multiple times by the same key in different arrangements. How much of a risk is this?

I don't think there is a regression. When we have a collation-requiring `Reduce`, we will necessarily create one reduce dataflow pipeline for each "type" of reduction; the duplication of the keys happens there no matter what, both in this PR and in `main`. What this PR changes is swapping out the subsequent DD `reduce` used to implement the `ReducePlan::Collation`, which requires two arrangements. It is replaced by a join, which .. has the potential to be a strict improvement: it requires no additional arrangements.

It may result in one "new" arrangement if the output of the collated reduce does need to be arranged, but in that case it should be the same size as the collated reduce's output arrangement, and we save the input arrangement.

The only regression I can think of is "now that it is a join, what if we choose some crazy new plan because joins are weird", and 100% I don't know if that will be regression-free or not.

ggevay commented:
> what if we choose some crazy new plan because joins are weird

These joins will be on unique keys though, which at least ensures that we won't have such blowups where some intermediate collection is much bigger than any of the inputs.

But the thing mentioned [here](https://github.com/MaterializeInc/materialize/pull/17013/files#r1676204776) is a regression, right? Because originally, the output arrangement of the Reduce would have been formed only after applying the Filter, if I understand things correctly?

ggevay commented on test/sqllogictest/explain/physical_plan_as_text.slt:
```diff
@@ -776,28 +776,79 @@ EXPLAIN PHYSICAL PLAN AS TEXT FOR
 MATERIALIZED VIEW collated_group_by_mv
 ----
 materialize.public.collated_group_by_mv:
-  Reduce::Collation
```
Could you adjust the comment at the top of this test?

Also, when we have a feature flag for the new transform, maybe we could have two copies of this test, where one copy is with the feature flag off and one is with on.

frankmcsherry replied:
Can we just delete the test instead? If there is no longer a `Reduce::Collated` to test, no need to keep the test.

ggevay replied:
Well, there will be, with the feature flag off.

What's the plan for `Reduce::Collated`? Will it be deleted (together with the feature flag) a few weeks from now when we have verified that no customers have any serious regressions?

ggevay commented on test/sqllogictest/attributes/mir_arity.slt:
```diff
@@ -96,9 +96,14 @@ EXPLAIN OPTIMIZED PLAN WITH(arity) AS TEXT FOR
+        Reduce group_by=[(#0 % 5)] aggregates=[sum((#0 * #1))] // { arity: 2 }
+          ReadStorage materialize.public.v // { arity: 2 }
+      ArrangeBy keys=[[#0]] // { arity: 2 }
+        Reduce group_by=[(#0 % 5)] aggregates=[max(#1)] // { arity: 2 }
```
Did the old version compute the key expression only once? If yes, then this is a possibility for a CPU regression.

We could prevent this by putting a complex key expression in a Map before each of the Reduces, in which case RelationCSE would grab also that Map.

frankmcsherry replied:
Probably! Though I think you could go either way on regression or not. Now, rather than doing all the keys and values for each reduction path, and then forking those into each reduction path, we "potentially" use less memory by creating the actual inputs directly, rather than sharing the work and then needing to clone the data to project it down.

ggevay replied:
Ok, then this is fine. In any case, CPU regressions are much less scary than memory regressions.

frankmcsherry commented:
> But the thing mentioned [here](https://github.com/MaterializeInc/materialize/pull/17013/files#r1676204776) is a regression, right? Because originally, the output arrangement of the Reduce would have been formed only after applying the Filter, if I understand things correctly?

I'm not sure it is. In `main` we would push the filter into the `ReducePlan::Collation` arrangement, where it could reduce the amount of memory required in the output arrangement. Now that arrangement doesn't even exist, and we have but miss out on the opportunity to push the filter down *even further* to the `SUM(#1)` reduction. The predicate would never make it that far on `main` at the moment.

frankmcsherry commented on test/sqllogictest/explain/physical_plan_as_text.slt:
```diff
@@ -851,49 +953,88 @@ MATERIALIZED VIEW collated_global_mv
 materialize.public.collated_global_mv:
```
Here's another test we could delete @ggevay

ggevay commented:
>> But the thing mentioned [here](https://github.com/MaterializeInc/materialize/pull/17013/files#r1676204776) is a regression, right? Because originally, the output arrangement of the Reduce would have been formed only after applying the Filter, if I understand things correctly?

> I'm not sure it is. In `main` we would push the filter into the `ReducePlan::Collation` arrangement, where it could reduce the amount of memory required in the output arrangement. Now that arrangement doesn't even exist, and we have but miss out on the opportunity to push the filter down *even further* to the `SUM(#1)` reduction. The predicate would never make it that far on `main` at the moment.

Ah yes. Great!

frankmcsherry commented:
The most recent commit is unfortunately line noise. When further shredding "basic" aggregates out they can no longer be sorted by reduction type, and we instead now sort by the reduction. It turns out that `AggregateFunc` sorts in roughly the opposite direction as `ReductionType`: all the `max` things now come first, and the accumulable reductions come second.

ggevay reviewed (commented):
The new change also looks good.

So, I think the only thing left to do here is adding a feature flag.

ggevay commented:
I kicked off a Nightly: https://buildkite.com/materialize/nightly/builds/8684

ggevay commented:
There are lots of Nightly failures:
- (There are a bunch of Testdrive failures in `mz-arrangement-sharing.td`, which should all go away after rewriting the test. ([Note that nowadays Testdrive also has a rewrite feature.](https://github.com/MaterializeInc/materialize/pull/26528)))
- The Parallel Workload failure seems unrelated, also happening on main.
- "Detect references to already closed issues" might be some problem with the test itself: maybe some GitHub API has changed. Could you please check, @MaterializeInc/testing ?
- The RQG failure is fixed [here](https://github.com/MaterializeInc/materialize/pull/28206) on main. Could you please rebase, @frankmcsherry ?
- I'm not sure what's going on with the upgrade checks failures, but maybe a rebase will fix those too?

nrainer-materialize commented:
>     * "Detect references to already closed issues" might be some problem with the test itself: maybe some GitHub API has changed. Could you please check, @MaterializeInc/testing ?

A rebase should fix this issue (fixed with https://github.com/MaterializeInc/materialize/pull/28336).

nrainer-materialize commented:
I think a rebase will be helpful here. Happy to assist if unexpected failures persist.

ggevay commented:
Kicked off a new Nightly: https://buildkite.com/materialize/nightly/builds/8868

ggevay commented:
[I think this is close to merging.](https://github.com/MaterializeInc/materialize/pull/17013#pullrequestreview-2191096866) Could you add a feature flag (on by default) and rebase? Then we'll run a fresh Nightly, and then hopefully merge.

ggevay commented:
New Nightly: https://buildkite.com/materialize/nightly/builds/10474

ggevay commented:
There is a new Testdrive failure, at `session.td:18`, because of the new feature flag.

antiguru commented:
Pushed an update to fix `sessions.td`.
